rfcs/33-apps_own_scheduling.md

```diff
@@ -0,0 +1,239 @@
+# Feature Name: `apps_own_scheduling`
+
+## Summary
+
+All scheduling information is attached directly to systems, and ultimately owned and configurable by the central app (or hand-written `bevy_ecs` equivalent).
+To ensure that this can be done safely, plugins specify invariants about the ordering of their systems which must be respected.
+Hard ordering constraints (where systems must be both ordered and separated by a hard sync point) are introduced as an assert-style protection to guard against a serious and common form of plugin misconfiguration.
+
+## Motivation
+
+The current plugin model works well in simple cases, but completely breaks down when the central consumer (typically, an `App`) needs to configure when and if their systems run. The main difficulties are:
+
+- 3rd party plugins can have "unfixable" ordering ambiguities or interop issues with other 3rd party plugins
+- plugins can not be added to states, and run criteria cannot be added
+- plugins can not be incorporated into apps and games that use a non-standard control flow architecture
+- end users who care about granular scheduling control for performance optimization have no way to change when systems run
+- worse, plugins can freely insert their own stages, causing sudden and uncontrollable bottlenecks
```

Comment:
Fair, the bottlenecks are unavoidable. However, I don't think that stages are needed for this.

When plugins need to have full world access, they can use exclusive systems, which can be ordered and visualized and reasoned about using the tools discussed here, without having to make strong assumptions about the top-level control flow of the app they're running in.

But plugins cannot and should not have access to the schedule itself. We've seen a huge proliferation of valid [high-level control flow patterns](https://github.com/bevyengine/bevy/discussions/2801#discussioncomment-1330209). Allowing plugins to insert systems directly into stages or add their own stages is impossible because these patterns each have their own incompatible patterns.